Improve wxBitmapBundle scaling logic and ensure it's consistent for files and MSW resources

vadz opened the pull request:
This is related to #22449, please see the discussion there for the full context, but the brief summary is that we now do our utmost to only upscale bitmaps by integer scale factors, as this is the only thing which works reasonably well (i.e. not horribly) in practice.

Please also check the new unit tests that describe in a hopefully relatively clear the expected behaviour (note that many of them would have failed without the rest of the changes here).

vadz committed:
Refactor wxBitmapBundleImplSet::GetPreferredBitmapSizeAtScale()

Extract the logic determining the scale to use in a reusable
DoGetPreferredSize() function to allow reusing it in other places.

There are no real changes here, this commit just moved the existing
code to the new function, but because it also changed it from using
wxSize to double, even git --color-moved doesn't show it as an actual
move.

vadz committed:
Don't use generated bitmaps when looking for preferred scale

The already scaled bitmaps shouldn't be used for deciding to which size
we can scale the bitmaps with good results, as scaling them again would
definitely be a bad idea.

vadz committed:
Use the same preferred size algorithm in wxBitmapBundleImplRC

Reuse DoGetPreferredSize() in this implementation in order to ensure
that we do the same thing when using bitmaps from disk files with
wxBitmapBundleImplSet or bitmaps from MSW resources with this one.

vadz committed:
Upscale images by integer factors only

Doing anything else results in really poor results, so don't even try
scaling images by 2.25, for example -- scaling by 2 looks much better
and the size difference is relatively insignificant and definitely not
worth the difference in quality.

vadz committed:
Break the tie in favour of smaller bitmaps at 1.5x factor

Prefer to use 1x bitmap to 2x one at 150% scaling, as it seems better to
use smaller (but still readable) bitmaps than overlarge ones in this
case.

vadz committed:
Extend wxBitmapBundle unit tests for bitmap scaling

Check not only that the resulting bitmap has the right size, but also
that it was obtained by scaling the correct bitmap.

Define the required machinery to allow using CATCH CHECK_THAT() macro to
succinctly verify both.

Note that this doesn't change the existing tests, but just rewrites them
in a new form and also adds more tests for a bundle with 3 bitmaps.

vadz committed:
Choose the best bitmap to rescale to the target size

It's not necessarily the largest bitmap, but one which may be scaled
using an integer factor.

Fix the tests to expect the correct results, now that they actually
pass.

vadz committed:
Access availableScales only sequentially in DoGetPreferredSize()

No real changes, just refactor the code to only access the array
elements consecutively in this function, in preparation to changing its
API.

vadz committed:
Change DoGetPreferredSize() to use a callback function

Instead of taking an array of scales, call GetNextAvailableScale() to
get them.

This allows centralising the logic for returning the available scales in
a single place, where it will be reused in the upcoming commits.

vadz committed:
Add wxBitmapBundleImpl::GetIndexToUpscale()

Ensure that wxBitmapBundleImplSet and wxBitmapBundleImplRC use the same
logic for actually selecting the bitmap to upscale, and not just for
deciding the size that it must have, too.

No real changes, but this should make impossible for these functions to
diverge once again -- and also make it simpler to reuse the same logic
in any other wxBitmapBundleImpl-derived classes in the future.

vadz committed:
Use non-integer scale that is exact multiple of an available one

Always rounding non-integer scale when upscaling is wrong, as we could
be able to upscale an existing x1.5 image to e.g. 300% DPI scaling with
relatively good results, so only do it if there is no available scale
that exactly divides the requested one.

Add a previously failing unit test which passes now.

vadz committed:
Use integer upscaling only in wxBitmapBundleImplArt

This still won't work perfectly for all existing custom bitmap
providers, but it shouldn't result in anything really catastrophic and
should help things for the providers that can return either a normal or
high resolution (i.e. x2) version of the same image.

vadz committed:
Use correct default size in wxArtProvider::GetBitmapBundle()

The returned bundle must have the specified default size, it's a
postcondition of this function, but the bitmap created by the art
provider does not have to be of the same size, so only use it if the
desired size was not explicitly specified.

vadz committed:
Remove unnecessary code from wxArtProvider::CreateBitmap()

Derived art providers don't need to scale the returned bitmaps as this
is already done by the base class GetBitmap() anyhow if necessary.

There should be no changes in behaviour, except that we may be using our
own bitmap scaling instead of GTK in some cases now.

vadz committed:
Remove useless CreateFromStdIcon() from wxMSW art provider

This function has become completely trivial after the previous commit,
so just remove it and create the bitmap from the icon directly in
CreateBitmap().

vadz committed:
Fix confusion between width and height in wxArtProvider

Amazingly, the fact that the bitmap height was compared with the
required width went unnoticed even though it was there since this code
was added back in e53a95bcb1 (Applied patch for ArtProvider.,
2005-03-12) and even survived migration to a different file in
2b7e668221 (Move bitmap resizing from wxDefaultArtProvider to base
class, 2022-02-05).

That this has never been a problem seems to conclusively prove that
non-square icons simply don't exist, but still fix the comparisons just
to be tidy.

vadz committed:
Allow upscaling wxArtProvider bitmaps by integer scale factor

While upscaling the bitmaps by factors smaller than 2 results in truly
awful results, upscaling them by 2, or 3, is better on high DPI screens
than using tiny bitmaps that are difficult to read in high resolution,
so do scale them up in this case.

This finally makes wxBitmapBundleImplArt::GetBitmap() work as expected,
as it now returns the upscaled bitmap rather than the original small
bitmap in the middle of a bigger canvas.

vadz commented:
@vslavik You're going to laugh (well, I hope you aren't going to cry, which seems to be the only other sane reaction to my vacillations), but I think I do finally have my preferred solution to the `wxBitmapBundleImplArt` problem. It consists of several pieces, _all_ of which are needed:

1. As previously mentioned, use integer upscaling only (e9b3fdaf278728cb3cdc769e37dd8878cfe99428).
2. Use the bitmap size but *iff* no explicit size was specified (0bd2a6c10552026fbdc249d11a9833fa5c961420).
3. Actually upscale bitmaps returned from `GetBitmap()` (f56b8a99a6283ad1e4ebec1306ed03208e42f93a).

I believe that this results in the most reasonable behaviour and don't plan to make any other changes to this. Please let me know why I'm wrong... or I'm going to put this in 3.1.7 :-/

vadz committed:
Fix bug with wrong GetIndexToUpscale() return value

Only assign known good values of the index to "indexLast" as otherwise
we could end up returning an invalid index value from it if "indexBest"
was never set and the last index wasn't a valid one, as it happened in
wxBitmapBundleImplSet containing an original bitmap and a x2 scaled copy
of it, which shouldn't be used for further scaling.

Add a unit test checking that this bug is fixed now.

vadz committed:
Document some issues involved in adding high DPI support

Notably mention that wxToolBar::SetToolBitmapSize() shouldn't be used.

vslavik reviewed (commented):
Generally, this looks good to me - thanks a lot!

The biggest change that I can see in testing is that the breakages are different - a lot of very large bitmaps (because they didn't have scale factor set and the art provider returned them at 2x). That's good-ish, I think - it is very obvious where the issue is.

Less good is then I still see some non-integer scaling in Poedit - this is will the code updated to set the correct scale factor on its bitmaps:

1. Running at 150%, all toolbar icons are slightly corrupted (the art provider has 1.5x versions) - scaling at factor 0.67. No idea why yet.
2. Running at 175% (1.5x versions used), the previously observed problematic bitmap that is converted out of `wxBitmapBundleImplArt` to the default impl, is still problematic and non-integer-scaled at 1.15 factor. There still is *some* difference between the two impls and I couldn't yet figure out what is going on.

vslavik commented:
> Less good is then I still see some non-integer scaling in Poedit - this is will the code updated to set the correct scale factor on its bitmaps:

OK, both of these are related to me changing GetSize() to GetDIPSize() as per the review comment. So it seems I was wrong about that. But without doing that, I don't know how should I return correctly sized bitmaps from wxArtProvider?

vadz commented on src/common/bmpbndl.cpp:
```diff
@@ -292,18 +292,35 @@ wxBitmap wxBitmapBundleImplSet::GetBitmap(const wxSize& size)
+
+        m_entries.push_back(entryNew);
+
+        return entryNew.bitmap;
```
The contract of `GetBitmap(size)` is that it will return the bitmap of exactly this size. So if it's called with a size which can't be obtained by integer scaling, it will scale the bitmap using non-integer factor. I think it might be actually better to do what `wxArtProvider::RescaleOrResizeIfNeeded()` does in this case, see https://github.com/wxWidgets/wxWidgets/blob/73d425b34285b1ba150a1748c2b9146d58125e09/src/common/artprov.cpp#L327,L333 but it will still do something to ensure that the bitmap has the requested size.

However normally only integer scaling should be used because only the bitmaps of "preferred" size should be requested. There are some exceptions, however. The one I've already encountered is that `wxToolBar` with a fixed tool bitmap size will always request bitmaps of this size. I'm not sure what to do about this, so far I've documented (in a a not yet pushed commit) that using `SetToolBitmapSize()` is just not recommended and that you should let toolbar adapt to the size of the bitmaps it has.

vadz commented on src/common/artprov.cpp:
```diff
@@ -428,8 +428,16 @@ wxBitmapBundle wxArtProvider::GetBitmapBundle(const wxArtID& id,
+                                id,
+                                client,
+                                size.IsFullySpecified() ? size
+                                                        : bitmap.GetSize()
```
Yes, thank you, you're right. But OTOH I don't expect many legacy art providers set any scale factor for the bitmaps they create, so this shouldn't matter.

But, again, I will still change this to `GetDIPSize()`.

vadz commented:
> > Less good is then I still see some non-integer scaling in Poedit - this is will the code updated to set the correct scale factor on its bitmaps:
> 
> OK, both of these are related to me changing GetSize() to GetDIPSize() as per the review comment. So it seems I was wrong about that.

No, I think you're right. But I don't understand why does this create problems in Poedit.

> But without doing that, I don't know how should I return correctly sized bitmaps from wxArtProvider?

I think any really DPI-aware applications should override `CreateBitmapBundle()` and not `CreateBitmap()` anyhow. E.g. you can't distinguish between being asked for a normal bitmap of size 32 and a high resolution version of a bitmap of size 16 in `CreateBitmap()`, but they may need to be different bitmaps.

And you also can't determine the appropriate size if `size == wxDefaultSize` because you don't have the window the bitmap will be used for, so you don't know its scale factor on a multi-monitor system.

vslavik commented on src/common/artprov.cpp:
```diff
@@ -165,42 +165,29 @@ class wxBitmapBundleImplArt : public wxBitmapBundleImpl
+        // ones should just override CreateBitmapBundle() directly), so we only
+        // return the original bitmap size, but hope that perhaps the provider
+        // will have a x2 version too, when our GetBitmap() is called.
+        return i++ ? 0.0 : 1.0;
```
I missed this in my initial review, but apparently was bothered by it subconsciously enough to wake up knowing it was the cause of my scaling issues. This shouldn't read "1.0", but should use the bitmap's scaling factor.

vadz replied:
Oops, yes, of course, thanks for finding this!

vslavik commented:
> No, I think you're right. But I don't understand why does this create problems in Poedit.

See my other comment, there were two places where the wrong size/scale was used.

> I think any really DPI-aware applications should override CreateBitmapBundle() and not CreateBitmap() anyhow.
…
> And you also can't determine the appropriate size if size == wxDefaultSize because you don't have the window the bitmap will be used for, so you don't know its scale factor on a multi-monitor system.

Yes, overriding CreateBitmapBundle is the "proper" way to upgrade to wx-3.2 and yes, any existing user code with HiDPI support will be suboptimal and full of compromises. But I got the impression from the other PR that you cared about such code too and want me to review with an eye towards it.

And with the one change suggested above, I really like this PR in that regard. The scaling issues are now easy to pinpoint (twice enlarged bitmaps vs subtly mis-scaled in different ways) and can be trivially fixed by setting the correct scaling factor. It makes it easy to upgrade Poedit (and I assume anybody doing any HiDPI work with wx will be in a similar situation) to 3.2's better way of doing things *incrementally* — quick fixes right away, then proper rewrite.

vadz committed:
Handle non-default scale factor in wxBitmapBundleImplArt

The existing code only worked correctly if overridden
wxArtProvider::CreateBitmap() returned bitmaps with scale factor of 1,
but not for anything else.

Account for the bitmap scale factor when determining the default size
and available scales correctly to fix this.

Co-authored-by: Václav Slavík <vaclav@slavik.io>

vadz commented:
> > No, I think you're right. But I don't understand why does this create problems in Poedit.
> 
> See my other comment, there were two places where the wrong size/scale was used.

I hope I've fixed them both in the latest commit, please let me know if I did anything wrong, thanks again!
 
> Yes, overriding CreateBitmapBundle is the "proper" way to upgrade to wx-3.2 and yes, any existing user code with HiDPI support will be suboptimal and full of compromises. But I got the impression from the other PR that you cared about such code too and want me to review with an eye towards it.

Yes, absolutely, I care about making things work as well as possible in this case, but I don't expect the impossible and I have probably misunderstood you as I thought you wanted to keep the existing code, overriding `CreateBitmap()` only.
 
> And with the one change suggested above, I really like this PR in that regard. The scaling issues are now easy to pinpoint (twice enlarged bitmaps vs subtly mis-scaled in different ways) and can be trivially fixed by setting the correct scaling factor. It makes it easy to upgrade Poedit (and I assume anybody doing any HiDPI work with wx will be in a similar situation) to 3.2's better way of doing things _incrementally_ — quick fixes right away, then proper rewrite.

Great to hear, thanks again for your help with it! And I'll merge this soon unless you (or anybody else, of course) see any problems with the last commit.

One thing I'm still not sure about is the tool bitmap size handling in `wxToolBar`. I think setting it may be quite common (even though it's totally unnecessary) and I wonder if saying "don't do this" is really the best we can do. Because in principle we could use integer scaling here too:

https://github.com/wxWidgets/wxWidgets/blob/73d425b34285b1ba150a1748c2b9146d58125e09/src/common/tbarbase.cpp#L480,L485

i.e. `sizeOrig*wxRound(scale)` instead of using `ToDIP()`. Do you see any reason _not_ to do this?

vslavik commented:
> Great to hear, thanks again for your help with it! And I'll merge this soon unless you (or anybody else, of course) see any problems with the last commit.

I don't, thanks!

> i.e. sizeOrig*int(scale) instead of using ToDIP(). Do you see any reason not to do this?

I'm more than a little confused by the code involved:

1. ~~The default size is 16x15 — how is that not causing problems, but other sizes are?~~ (because default size is minimum)
2. Why is the code explicitly allowing downscaling only in the very edge case of a) different tool sizes, b) with equal count of both, c) that doesn't happen to match the ideal size https://github.com/wxWidgets/wxWidgets/blob/731d29baceda04c5781c6e204cec3d9df8fe9d53/src/common/bmpbndl.cpp#L670-L677 I mean, why even bother?
3. Why is it feeding the determined size back as the ideal size next time this function runs? Should the input to GetConsensusSizeFor be `m_requestedBitmapSize` only?

I would make a different choice in the snippet above - in a tie, I would pick the size closer to `sizeIdeal`, instead of the larger one.

As for using `sizeOrig*int(scale)` instead of `ToDIP()`, unless I'm missing something it would be incorrect:

1. `sizeOrig` (aka m_defaultWidth/m_defaultHeight) is expressed in physical pixels
2. `ToDIP(sizeOrig)` converts it to DIP
5. In `GetConsensusSizeFor` it is immediately converted back by multiplying it by `win->GetDPIScaleFactor()`
6. (…and isn't actually used for much)

In any case, it wouldn't have any effect except in an edge case. The actual cause of the problem is further down *after* obtaining the consensus, I think:
https://github.com/wxWidgets/wxWidgets/blob/73d425b34285b1ba150a1748c2b9146d58125e09/src/common/tbarbase.cpp#L497-L500

You're right that calling `SetToolBitmapSize()` may be common. Unless I misunderstand the problem, I would:

1. Outright deprecate `SetToolBitmapSize()`
2. Either ignore the value provided or only use it as hint for `GetConsensusSizeFor`, but don't `IncTo` to it (possibly only if scaling factor > 1.0)
3. When adjusting the size, use resize, not rescale — I'm not sure if it is feasible, if this is done in win32, probably not

vadz commented:
Thanks, I'm going to merge this one because it already has so many commits/changes, but I'll create another PR addressing the toolbar bitmap size issues and taking your comments into account a.s.a.p.

For the record, at least parts of this code are needed in order to make the toolbar sample, which illustrates showing bitmaps of different sizes, work correctly when moving it between the displays using different DPIs. I think that simplifying it too much would break it again (it was already broken more than once...), but I'll test it to be certain.

And thanks again for the review/discussion!
